### PR TITLE
OAK-9213 - Support feature vector similarity / image similarity in Oak ES

### PR DESCRIPTION
**Pull request by averma21**

### CODE REVIEW

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java [hunk 1]
```diff
@@ -81,6 +91,11 @@ void addProperty(String fieldName, Object value) {
         properties.put(fieldName, value);
     }
 
+    void addSimField(String name, Blob value) throws IOException{
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java]: Could you rename the method to "addSimilarityField"? I first thought it means "addSimulatedField" :-)

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java [hunk 2]
```diff
@@ -148,4 +166,19 @@ public String build() {
...
     }
+
+    public static Collection<Field> newSimilarityFields(String name, Blob value) throws IOException {
+        Collection<Field> fields = new ArrayList<>(1);
+        byte[] bytes = new BlobByteSource(value).read();
+        fields.add(newSimilarityField(name, bytes));
+        return fields;
+    }
+
+    private static Field newSimilarityField(String name, byte[] bytes) {
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java]: For my taste, one of the methods newSimilarityField isn't needed... instead I would inline the logic. That simplifies the code a tiny bit.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java [hunk 3]
```diff
@@ -181,6 +181,10 @@ protected void indexSimilarityBinaries(ElasticDocument doc, PropertyDefinition p
         // TODO : not implemented
         // see https://www.elastic.co/blog/text-similarity-search-with-vectors-in-elasticsearch
         // see https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html
+
+        doc.addSimField(pd.name, blob);
+
+        // TODO : RERANK
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java]: This TODO I don't understand... I thought we only rerank (in Lucene) when querying?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 4]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+            String [] pathTokens = text.substring(1).split("/");
+            NodeState targetNodeState = rootState;
+            for (String token : pathTokens) {
+                targetNodeState = targetNodeState.getChildNode(token);
+            }
+            if (!targetNodeState.exists()) {
+                throw new IllegalArgumentException("Could not find node " + text);
+            }
+            for (PropertyDefinition pd : sp) {
+                int slashLastIndex = pd.name.lastIndexOf('/');
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: Could you use the Oak PathUtils instead? I think it is safer, _and_ simpler.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 5]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+                String propertyPath = slashLastIndex != -1 ? pd.name.substring(0, slashLastIndex) : "";
+                String propertyName = pd.name.substring(slashLastIndex+1);
+                String [] propertyPathTokens = propertyPath.split("/");
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: use PathUtils instead here as well

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 6]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+                for (String token : propertyPathTokens) {
+                    if (token.isEmpty()) {
+                        break;
+                    }
+                    tempState = tempState.getChildNode(token);
+                }
+                PropertyState ps = tempState.getProperty(propertyName);
+                Blob property = ps != null ? ps.getValue(Type.BINARY) : null;
+                if (property == null) {
+                    LOG.error("Couldn't find property {} on {}", pd.name, text);
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: I would probably log a warning here instead, as it's not an extremely sever problem.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 7]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+                    continue;
+                }
+                byte[] bytes = null;
+                try {
+                    bytes = new BlobByteSource(property).read();
+                } catch (IOException e) {
+                    LOG.error("Error reading bytes from property {} on {}", pd.name, text);
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: Here, error makes sense, because IOException would mean something wrong with the repository (e.g. Blob not found).

But the exception stack trace also needs to be logged!

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java [hunk 8]
```diff
@@ -46,4 +49,52 @@ public static String idFromPath(@NotNull String path) {
...
+    /**
+     * Converts a given byte array (of doubles) to a list of doubles
+     * @param array given byte array
+     * @return list of doubles
+     */
+    public static List<Double> toDoubles(byte[] array) {
+        int blockSize = Double.SIZE / Byte.SIZE;
+        ByteBuffer wrap = ByteBuffer.wrap(array);
+        int capacity = array.length / blockSize;
+        List<Double> doubles = new ArrayList<>(capacity);
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java:69]: I think it would make sense to log a warning if the size of the array is unexpected... e.g. size 1, or size 7, or so.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java [hunk 9]
```diff
@@ -46,4 +49,52 @@ public static String idFromPath(@NotNull String path) {
...
+    /**
+     * Converts a given list of double values into a byte array
+     * @param values given list of doubles
+     * @return byte array
+     */
+    public static byte[] toByteArray(List<Double> values) {
+        int blockSize = Double.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[values.size() * blockSize];
+        for (int i = 0, j = 0; i < values.size(); i++, j += blockSize) {
+            ByteBuffer.wrap(bytes, j, blockSize).putDouble(values.get(i));
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java]: I think it's better to only call ByteBuffer.wrap once (outside of the loop).

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java [hunk 10]
```diff
@@ -154,4 +173,51 @@ public void propertyExistenceQuery() throws Exception {
                 Arrays.asList("/test/a", "/test/b")));
     }
 
+
+    @Test
+    public void testSim() throws Exception {
```
Comment (thomasmueller) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java]: testSimilarity

**Review by thomasmueller — CHANGES_REQUESTED** [review]

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java [hunk 11]
```diff
@@ -202,6 +206,13 @@ private static void mapIndexRules(ElasticIndexDefinition indexDefinition, XConte
                 }
             }
             mappingBuilder.endObject();
+
+            if (useInSimilarity) {
```
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:215]: sim:, dense_vector can be moved to some constant. And dims value = 1024 should be made configurable because this can change based on the underlying implementation by dam (which is not in our control)

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java [hunk 12]
```diff
@@ -96,16 +96,16 @@ protected IndexNode acquireIndexNode(String indexPath) {
...
+        return Strings.toString(new ElasticRequestHandler(plan, getPlanResult(plan), rootState).baseQuery());
     }
 
     @Override
     public Cursor query(IndexPlan plan, NodeState rootState) {
         final Filter filter = plan.getFilter();
         final FulltextIndexPlanner.PlanResult planResult = getPlanResult(plan);
 
-        final ElasticRequestHandler requestHandler = new ElasticRequestHandler(plan, planResult);
+        final ElasticRequestHandler requestHandler = new ElasticRequestHandler(plan, planResult, rootState);
```
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java:108]: I assume this rootState is to get the vector values from the image node since the similarity query would only have the path to the image node.

If we want to avoid changing this signature  - what we can do is call ES twice - first time get the node data for the path in query, read the value for vector dims and then use them in the similarity query

**Review by nit0906 — COMMENTED** [review]
Added some comments

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java (hunk 3 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java]: +1

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java [hunk 13]
```diff
@@ -181,6 +181,10 @@ protected void indexSimilarityBinaries(ElasticDocument doc, PropertyDefinition p
         // TODO : not implemented
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java]: we can remove this

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java (hunk 2 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java]: I don't think both `newSimilarityField ` methods are used. Can we remove them?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 11 above)
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:215]: we need to check how this affects the index size. Since this is used for a specific feature, it could be smart to configure this as nested doc (that underneath creates a separate index). I would keep this for a separate PR anyway.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 14]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+                    }
+                    tempState = tempState.getChildNode(token);
+                }
+                PropertyState ps = tempState.getProperty(propertyName);
+                Blob property = ps != null ? ps.getValue(Type.BINARY) : null;
+                if (property == null) {
+                    LOG.error("Couldn't find property {} on {}", pd.name, text);
+                    continue;
+                }
+                byte[] bytes = null;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: minor: `null` is redundant here

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 15]
```diff
@@ -150,11 +165,20 @@ public BoolQueryBuilder baseQuery() {
...
+                if (sp.isEmpty()) {
+                    // SimilarityImpl in oak-core sets property restriction for sim search and the query is something like
+                    // mlt?mlt.fl=:path&mlt.mindf=0&stream.body=<path> . We need parse this query string and turn into a query
+                    // elastic can understand.
+                    String mltQueryString = propertyRestrictionQuery.replace("mlt?", "");
+                    boolQuery.must(moreLikeThisQuery(mltQueryString));
+                } else {
+                    String mltQueryString = propertyRestrictionQuery.replace("mlt?", "");
+                    boolQuery.must(similarityQuery(mltQueryString, sp));
+                }
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]:
```suggestion
                String mltQueryString = propertyRestrictionQuery.replace("mlt?", "");
                if (sp.isEmpty()) {
                    // SimilarityImpl in oak-core sets property restriction for sim search and the query is something like
                    // mlt?mlt.fl=:path&mlt.mindf=0&stream.body=<path> . We need parse this query string and turn into a query
                    // elastic can understand.
                    boolQuery.must(moreLikeThisQuery(mltQueryString));
                } else {
                    boolQuery.must(similarityQuery(mltQueryString, sp));
                }
```

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 16]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+        for (String param : mltQueryString.split("&")) {
+            String[] keyValuePair = param.split("=");
+            if (keyValuePair.length != 2 || keyValuePair[0] == null || keyValuePair[1] == null) {
+                throw new RuntimeException("Unparsable native query for fv similarity: " + mltQueryString);
+            } else {
+                if ("stream.body".equals(keyValuePair[0])) {
+                    text = keyValuePair[1];
+                    break;
+                }
+            }
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: this is a duplicate of `MoreLikeThisHelperUtil.getParamMapFromMltQuery` already used in `moreLikeThisQuery`. Please reuse the same method.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 17]
```diff
@@ -263,6 +287,69 @@ private boolean requiresFacets() {
...
+                    LOG.error("Error reading bytes from property {} on {}", pd.name, text);
+                    continue;
+                }
+                String similarityPropFieldName = FieldNames.createSimilarityFieldName(pd.name);
+                Map<String, Object> paramMap = new HashMap<>();
+                paramMap.put("query_vector", toDoubles(bytes));
+                paramMap.put("field_name", similarityPropFieldName);
+                ScriptScoreQueryBuilder scriptScoreQueryBuilder = new ScriptScoreQueryBuilder(new ExistsQueryBuilder(similarityPropFieldName),
+                        new Script(ScriptType.INLINE, Script.DEFAULT_SCRIPT_LANG, "cosineSimilarity(params.query_vector, params.field_name) + 1.0",
+                                Collections.emptyMap(), paramMap));
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]:
```suggestion
                ScriptScoreQueryBuilder scriptScoreQueryBuilder = scriptScoreQuery(existsQuery(similarityPropFieldName), 
                        new Script(ScriptType.INLINE, Script.DEFAULT_SCRIPT_LANG, "cosineSimilarity(params.query_vector, params.field_name) + 1.0",
                        Collections.emptyMap(), paramMap));
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java [hunk 18]
```diff
@@ -16,16 +16,35 @@
...
 import org.apache.jackrabbit.oak.api.Tree;
+import org.apache.jackrabbit.oak.api.Type;
 import org.apache.jackrabbit.oak.plugins.index.search.util.IndexDefinitionBuilder;
 import org.junit.Test;
 
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java]:
```suggestion
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java [hunk 19]
```diff
@@ -16,16 +16,35 @@
...
+import java.nio.charset.Charset;
+import java.util.ArrayList;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java]:
```suggestion
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java [hunk 20]
```diff
@@ -154,4 +173,51 @@ public void propertyExistenceQuery() throws Exception {
...
+            while (result.hasNext()) {
+                String next = result.next();
+                current.add(next);
+            }
+            assertNotEquals(baseline, current);
+            baseline.clear();
+            baseline.addAll(current);
+        }
+
+        Thread.sleep(10000);
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java]: can this be removed?

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java (hunk 10 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest.java]: +1 on changing the name. I tend to not include `test` since we already have the annotation telling us this is a test. I think that something like `vectorSimilarity` is nicer.

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java (hunk 3 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocumentMaker.java]: This was being used for brute force reordering by calculating distance b/w feature vectors after results from lucene were returned. Currently we are are using brute force approach in ES itself, so we don't need this reordering for now. However, I have removed this todo comment.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 11 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:215]: Done.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java (hunk 12 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java:108]: We discussed this earlier in scrum and concluded that we should avoid unnecessary remote calls.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java (hunk 8 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java:69]: Added a warning above. Please verify if that's what you meant.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java [hunk 21]
```diff
@@ -148,4 +166,11 @@ public String build() {
     public String toString() {
         return build();
     }
+
+    public static Collection<Field> newSimilarityFields(String name, Blob value) throws IOException {
```
Comment (fabriziofortino) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticDocument.java]: I don't think this method is used. Can we remove it with all the import references on Lucene?

**Review by fabriziofortino — APPROVED** [review]

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java [hunk 22]
```diff
@@ -202,6 +211,14 @@ private static void mapIndexRules(ElasticIndexDefinition indexDefinition, XConte
...
+                mappingBuilder.startObject(FieldNames.createSimilarityFieldName(name));
+                mappingBuilder.field("type", ES_DENSE_VECTOR_TYPE);
+                mappingBuilder.field(ES_DENSE_VECTOR_DIM_PROP, Integer.parseInt(System.getProperty(PROP_ES_SIMILARITY_SEARCH_DENSE_VECTOR_SIZE,
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java]: Hm, this is using System.getProperty... I think it would be better if this is configurable in the index definition itself... Would that be possible?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 23]
```diff
@@ -263,6 +288,57 @@ private boolean requiresFacets() {
...
     }
 
+    private QueryBuilder similarityQuery(String mltQueryString, List<PropertyDefinition> sp) {
+        LOG.debug("parsing similarity query on {}", mltQueryString);
+        Map<String, String> mltParamMap = MoreLikeThisHelperUtil.getParamMapFromMltQuery(mltQueryString);
+        String text = mltParamMap.get(MoreLikeThisHelperUtil.MLT_STREAM_BODY);
+        BoolQueryBuilder query = boolQuery();
+        if (text != null && !sp.isEmpty()) {
+            LOG.debug("generating similarity query for {}", text);
+            String [] pathTokens = text.substring(1).split("/");
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: It should be possible to use PathUtils.elements here, instead of String.split

**Review by thomasmueller — CHANGES_REQUESTED** [review]

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 22 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java]: I think vector representation would be obtained from an external service and it would provide same length of vectors for all images. So duplicating this information for each index doesn't make sense currently. If such a need arises in future, it can be done without breaking backwards compatibility. WDYT?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java (hunk 12 above)
Comment (nit0906) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java:108]: ok

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java [hunk 24]
```diff
@@ -169,14 +169,13 @@ public BoolQueryBuilder baseQuery() {
                 for (IndexDefinition.IndexingRule r : elasticIndexDefinition.getDefinedRules()) {
                     sp.addAll(r.getSimilarityProperties());
                 }
+                String mltQueryString = propertyRestrictionQuery.replace("mlt?", "");
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java]: Hm, this replaces all occurrences... So if the query is "mtl?abcmtl?" then the result is "abc" instead of "abcmtl?". So I think we should use

    propertyRestrictionQuery.replaceFirst("^mtl\\?", "");

Or the faster (simpler?):

    propertyRestrictionQuery.substring("mtl?".length());

**Review by thomasmueller — CHANGES_REQUESTED** [review]

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 22 above)
Comment (thomasmueller) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java]: In my view, there's likely only one index that uses image similarity... Yes I think it makes sense to specify this in the index definition. Sure we can do it later if we are in a rush, but I would probably do it now...

**Review by thomasmueller — CHANGES_REQUESTED** [review]

**Comment by averma21** [comment]
Committed in - http://svn.apache.org/viewvc?view=revision&revision=1882970
